### ISSUE TITLE
How long to reproduce the result of Basket option?

### ISSUE THREAD

**Issue by xwuShirley** [juliusberner/deep_kolmogorov#1]
Dear authors @juliusberner ,

Very nice work. Thank you for the code. 

We are able to reproduce your result for the Black-Scholes model (Table 1),  heat equation with paraboloid initial condition (Table 3), and Gaussian initial condition (Table 4). However, we have a problem with the Basket option (Table 2). The time for gradient step 4k is not within 811s as the table suggested. In fact, we didn't even get any results even though we kept the code run for 12 hours.

We guess the main reason perhaps due to the two inner loops in that the code for the simulation of the solution (https://github.com/juliusberner/deep_kolmogorov/blob/abe752b4c9176f69dbf2d3d797f69bf5fba79859/deep_kolmogorov/pdes.py#L216)
Do you think we can remove this part and make it parallel?

I wonder if you already save some of the simulated data that we could directly use. Thank you very much!

**Comment by juliusberner** [comment]
Dear @xwuShirley,
thank you very much for your interest. 

The avg. time presented in our tables refers to the training time only, see also Item 1. below. For our examples with explicit solutions evaluation is very fast. However, for the Basket option, we lack an explicit solution and we compute one batch of 1e17 evaluation samples each simulated using 1e20 Euler-Maruyama (EM) realizations with 25 steps. In order to speed up the computations we distributed the batch to 4 Tesla V100s, but this remains computationally expensive, see also [footnote [10]](https://arxiv.org/pdf/2011.04602.pdf#Hfootnote.9) in the appendix of our paper.

**Solutions:**

1. We just computed the intermediate errors (at step 0, 4k, 8k, ...) in order to showcase the convergence of our algorithm. To just train the model, one can set `n_test_batches` to 0 and then use the checkpoint to evaluate the trained model afterwards.

2. One can try to parallelize the EM simulations, i.e. the `solution` method in the snipped you linked. However, I have already been profiling different methods (e.g. I tried to transform the for-loop into computations on a new axis using `torch.einsum`) and ran either into out-of-memory issues or could not improve the overall runtime.

3. Of course, one could also precompute evaluation samples. However, we decided to simulate them on-the-fly for the following reasons: 1.) If one wants to change the parameter domains of the Basket option, one would need to recompute them 2.) we would need to provide the samples to the user and this needs a considerable amount of storage 3.) we wanted to have independent evaluation samples at each step 0, 4k, 8k, ... as it is also the case for the other PDEs and 4.) if we precompute them, one needs to pay attention to not allow the user to use the same seed for training, as this could potentially train and evaluate on the same points. However, if you really need precomputed evaluation data, I could simulate it and send it to you - Note, that you would then need to change the dataloader accordingly.

I hope, I could clarify your question. Do not hesitate to contact me, if you have any further questions, if you have solutions to Item 2., if you need precomputed data, or if you want to discuss future research directions.

FYI: I am also working on an updated version of this repo with further functionality.

**Comment by xwuShirley** [comment]
Hi @juliusberner 

Thank you very much for your timely response and the details.
Thank you for the suggestion to compute the test error only at 0, 4k, etc. I will try it.
In regards to  Item 3, I now understand why you train the data and test the data on the fly. But since both train and test loader generate the data on-the-fly, do you think that the test_loader might contain a large portion of data from  train_loader, particularly when the training time is longer? The reason why test errors getting smaller perhaps due to the powerful property of the neural network -- memorization. People might be more interested in a model that can generalize to some new data that is not in the train_loader.  This is something that might be worth checking.

But regarding Item  2,  I think we can still do the parallel for mini-batches? That is we let full_bs = len(batch["t"]) and we compute small_bs =full_bs //10 in parallel. Using this we would reduce  **full_bs //10** times!
Could you provide your torch.einsum code for me to try? 

Thank you very much!

**Comment by juliusberner** [comment]
Hi @xwuShirley,

you are welcome! :)

I think 0, 4k, etc. is even the default, so I would only compute the test error at the end of training, e.g. after step 28k.

The data of the `test_loader` and the `train_loader` is independent of each other, i.e. (neglecting floating-point arithmetic and pseudorandom number generators) the probability that they contain the same data point is zero.

Unfortunately, I did not save the code using `torch.einstein` as it did not improve performance. Parallelizing along the batch-axis is already implemented for multiple GPUs. However, parallelizing on a single GPU (e.g. in `n` mini-batches as you suggested) will only work if your GPU memory can fit `n`-times the 1e20 Euler-Maruyama computations, which was not the case when I tried it.

**Comment by xwuShirley** [comment]
@juliusberner  
Thank you very much for the explanation on the generalization part. I tried to match the data of train and test for Black-Scholes option. It indeed gives an empty intersection set. 

For the basket part. I think I have reproduced your results -- what I did is to generate only 6k test data and then save it. I rewrote the test_

    def _test_loop(self):
        # test
        self.model.eval()
        # zero running metrics
        self.test_metr.zero()
        with torch.no_grad(): 
            batch = torch.load(test_dir+'/test.pt')
            gt_load = torch.load(test_dir+'/test_solution.pt') 
            ... 
Thank you very much!
